storage/range.go

```diff
@@ -1677,20 +1733,66 @@ func (r *Range) Append(entries []raftpb.Entry) error {
 
 // ApplySnapshot implements the multiraft.WriteableGroupStorage interface.
 func (r *Range) ApplySnapshot(snap raftpb.Snapshot) error {
-	// TODO(bdarnell): discard any existing log that is obsoleted by this snapshot.
-	desc := &proto.RangeDescriptor{}
-	err := gogoproto.Unmarshal(snap.Data, desc)
+	snapData := proto.RaftSnapshotData{}
+	err := gogoproto.Unmarshal(snap.Data, &snapData)
 	if err != nil {
+		return nil
+	}
+
+	hardStateKey := engine.RaftHardStateKey(r.Desc().RaftID)
+	encodedHardStateKey := engine.MVCCEncodeKey(hardStateKey)
+	descKey := engine.RangeDescriptorKey(r.Desc().StartKey)
+	var desc proto.RangeDescriptor
+	seenDesc := false
+	batch := engine.NewBatch(r.rm.Engine())
+
+	// Delete everything in the range and recreate it from the snapshot.
+	for iter := newRangeDataIterator(r, r.rm.Engine()); iter.Valid(); iter.Next() {
+		if bytes.Equal(encodedHardStateKey, iter.Key()) {
+			// Don't replace the raft HardState.
+			continue
+		}
+		if err := batch.Clear(iter.Key()); err != nil {
+			return err
+		}
+	}
+
+	for _, kv := range snapData.KV {
+		key, _, isValue := engine.MVCCDecodeKey(kv.Key)
+		switch {
+		case bytes.Equal(descKey, key) && isValue && !seenDesc:
+			// Save the descriptor so we can assign it to r.Desc.
```

Comment:
Let's write everything in one simple loop and then read the range descriptor afterwards.

I'm realizing here that there's a nasty bug in how we're using MVCCScan(consistent=false) when starting up. We could end up skipping an intent which had actually been committed. What I'm going to do is make resolving the intents part of the commit trigger on split / merge / update replicas, as in these cases the intents are guaranteed to be on the same range as the end transaction. This will allow you to call MVCCGet(consistent=false) correctly to get the correct value for the range descriptor. If there is an intent, it will be one for which an EndTransaction hasn't yet arrived.